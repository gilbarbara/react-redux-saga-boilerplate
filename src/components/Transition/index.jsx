import React from 'react';
import PropTypes from 'prop-types';
import { CSSTransition, TransitionGroup } from 'react-transition-group';

import transitions from './transitions';

const Transition = ({ children, className, style, transition, ...rest }) => {
  const Component = transitions[transition];

  if (!Component) {
    console.error(`Invalid transition: ${transition}`); //eslint-disable-line no-console
    return null;
  }

  return (
    <TransitionGroup className={className} style={style}>
      {React.Children.map(children, child => (
        <CSSTransition classNames={transition} {...rest}>
          <Component>{child}</Component>
        </CSSTransition>
      ))}
    </TransitionGroup>
  );
};

Transition.propTypes = {
  appear: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  enter: PropTypes.bool,
  exit: PropTypes.bool,
  style: PropTypes.object,
  timeout: PropTypes.number,
  transition: PropTypes.oneOf(Object.keys(transitions)),
};

Transition.defaultProps = {
  appear: false,
  enter: true,
  exit: true,
  style: null,
  timeout: 300,
  transition: 'fade',
};

export default Transition;
