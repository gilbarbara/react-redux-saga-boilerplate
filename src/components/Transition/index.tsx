import { Children, CSSProperties, ReactNode } from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';

import { Transitions } from '~/types';

import transitions, { classNames } from './transitions';

interface Props {
  appear: boolean;
  children: ReactNode;
  className?: string;
  enter: boolean;
  exit: boolean;
  style?: CSSProperties;
  timeout: number;
  transition: Transitions;
}

function Transition({ children, className, style, transition, ...rest }: Props) {
  const Component = transitions[transition];

  if (!Component) {
    console.error(`Invalid transition: ${transition}`); // eslint-disable-line no-console

    return null;
  }

  return (
    <TransitionGroup className={className} style={style}>
      {Children.toArray(children)
        .filter(child => !!child)
        .map((child, key) => (
          // eslint-disable-next-line react/no-array-index-key
          <CSSTransition key={key} classNames={classNames[transition]} {...rest}>
            <Component>{child}</Component>
          </CSSTransition>
        ))}
    </TransitionGroup>
  );
}

Transition.defaultProps = {
  appear: true,
  enter: true,
  exit: true,
  timeout: 300,
  transition: 'fade',
};

export default Transition;
