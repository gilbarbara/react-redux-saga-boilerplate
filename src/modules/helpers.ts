import { Reducer } from 'redux';
import { format } from 'date-fns';
import produce from 'immer';
import is from 'is-lite';

import {
  ActionCreator,
  ActionsMapReducer,
  GenericFunction,
  LoggerOptions,
  PlainObject,
  SortFunction,
  StoreAction,
} from 'types';

/**
 * Create an action
 */
export function createAction<T extends GenericFunction>(
  type: string,
  payloadCreator: T,
): ActionCreator<Parameters<T>, ReturnType<T>> {
  if (!payloadCreator) {
    throw new TypeError('Expected a function');
  }

  return (...args: any[]) => ({
    type,
    payload: payloadCreator(...args),
  });
}

/**
 * Create a reducer
 */
export function createReducer<State>(
  actionsMap: ActionsMapReducer<State>,
  defaultState: State,
): Reducer<State, StoreAction> {
  return (state = defaultState, action: StoreAction) =>
    produce(state, (draft: State) => {
      const fn = actionsMap[action.type];

      if (fn) {
        return fn(draft, action);
      }

      return draft;
    });
}

/**
 * Async Sleep
 */
export function sleep(seconds = 1) {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
}

/**
 * Get unix timestamp in seconds
 */
export function getUnixtime(): number {
  return Math.floor(Date.now() / 1000);
}

/**
 * Check if cache is valid
 */
export function hasValidCache(lastUpdated: number, max = 10): boolean {
  if (!navigator.onLine) {
    return true;
  }

  return lastUpdated + max * 60 > getUnixtime();
}

export function keyMirror<T extends PlainObject>(input: T): { [K in keyof T]: K } {
  if (!is.plainObject(input)) {
    throw new TypeError('Expected an object');
  }

  const output: any = {};

  for (const key in input) {
    /* istanbul ignore else */
    if (!Object.prototype.hasOwnProperty.call(output, key)) {
      output[key] = key;
    }
  }

  return output;
}

/**
 * Log grouped messages to the console
 */
export function logger(
  type: string,
  title: string,
  data: unknown,
  options: LoggerOptions = {},
): void {
  /* eslint-disable no-console */
  const { collapsed = true, hideTimestamp = false, skip = false, typeColor = 'gray' } = options;
  const groupMethod = collapsed ? console.groupCollapsed : console.group;

  /* istanbul ignore else */
  if (process.env.NODE_ENV === 'development') {
    const parts = [`%c ${type}`];
    const styles = [`color: ${typeColor}; font-weight: lighter;`, 'color: inherit;'];
    const time = format(new Date(), 'HH:mm:ss');

    if (!hideTimestamp) {
      styles.push('color: gray; font-weight: lighter;');
    }

    parts.push(`%c${title}`);

    if (!hideTimestamp) {
      parts.push(`%c@ ${time}`);
    }

    /* istanbul ignore else */
    if (!skip && !window.HIDE_LOGS) {
      groupMethod(parts.join(' '), ...styles);
      console.log(data);
      console.groupEnd();
    }
  }
  /* eslint-enable no-console */
}

/**
 * Sort an array with localeCompare
 */
export function sortByLocaleCompare(
  key?: string,
  options: Intl.CollatorOptions & { descending?: boolean } = {},
): SortFunction {
  const { descending, ...compareOptions } = options;

  if (key) {
    if (descending) {
      return <T extends PlainObject>(left: T, right: T) =>
        right[key].toLowerCase().localeCompare(left[key].toLowerCase(), undefined, compareOptions);
    }

    return <T extends PlainObject>(left: T, right: T) =>
      left[key].toLowerCase().localeCompare(right[key].toLowerCase(), undefined, compareOptions);
  }

  if (descending) {
    return <T extends string>(left: T, right: T) =>
      right.toLowerCase().localeCompare(left.toLowerCase(), undefined, compareOptions);
  }

  return <T extends string>(left: T, right: T) =>
    left.toLowerCase().localeCompare(right.toLowerCase(), undefined, compareOptions);
}
