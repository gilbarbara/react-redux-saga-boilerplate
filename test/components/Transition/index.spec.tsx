import React from 'react';
import { noop } from '@gilbarbara/helpers';
import { render } from 'test-utils';

import Transition from '~/components/Transition/index';

describe('Transition', () => {
  beforeAll(() => {
    vi.spyOn(console, 'error').mockImplementation(noop);
  });

  afterAll(() => {
    vi.resetAllMocks();
  });

  it('should render properly', () => {
    const { container } = render(
      <Transition transition="fade">
        <div className="transition" />
      </Transition>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should not render if transition don't exist", () => {
    const { container } = render(
      // @ts-ignore
      <Transition transition="rotate">
        <div className="transition" />
      </Transition>,
    );

    expect(container).toMatchSnapshot();
  });
});
