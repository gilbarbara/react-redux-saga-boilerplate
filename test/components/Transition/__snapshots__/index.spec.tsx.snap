// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transition > should not render if transition don't exist 1`] = `<div />`;

exports[`Transition > should render properly 1`] = `
<div>
  <div>
    <div
      class="css-cqqmi6 fade-appear fade-appear-active"
    >
      <div
        class="transition"
      />
    </div>
  </div>
</div>
`;
